.gitlab-ci.yml

```diff
@@ -2,26 +2,30 @@ image: python:3.10
 
 stages:
   - deploy
-  - deploy-latest
+  - promote
 
-deploy_image:
+deploy-images:
   stage: deploy
   image:
     name: gcr.io/kaniko-project/executor:debug
     entrypoint: [""]
   script:
     - echo "{\"auths\":{\"$CI_REGISTRY\":{\"username\":\"$CI_REGISTRY_USER\",\"password\":\"$CI_REGISTRY_PASSWORD\"}}}" > /kaniko/.docker/config.json
-    - /kaniko/executor --context $CI_PROJECT_DIR --dockerfile $CI_PROJECT_DIR/Dockerfile --destination $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG
-  only:
-    - tags
+    - executor --context $CI_PROJECT_DIR --dockerfile $CI_PROJECT_DIR/Dockerfile --destination $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-alpine
+    - executor --context $CI_PROJECT_DIR --dockerfile $CI_PROJECT_DIR/Dockerfile --destination $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-slim --build-arg PYTHON_FLAVOR=slim
+  rules:
+    - if: $CI_COMMIT_TAG
 
-deploy-latest:
-  stage: deploy-latest
+tag-latest:
+  stage: promote
   image:
     name: gcr.io/go-containerregistry/crane:debug
     entrypoint: [""]
   script:
-    - mkdir /root/.docker && echo "{\"auths\":{\"$CI_REGISTRY\":{\"username\":\"$CI_REGISTRY_USER\",\"password\":\"$CI_REGISTRY_PASSWORD\"}}}" > /root/.docker/config.json
-    - /ko-app/crane cp $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG $CI_REGISTRY_IMAGE:latest
-  only:
-    - tags
+    - crane auth login -u $CI_REGISTRY_USER -p $CI_REGISTRY_PASSWORD $CI_REGISTRY
+    - crane tag $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-alpine ${CI_COMMIT_TAG} # /python-gitlab:v1.2.3
+    - crane tag $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-alpine latest           # /python-gitlab:latest
+    - crane tag $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-alpine alpine           # /python-gitlab:alpine
+    - crane tag $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-slim slim               # /python-gitlab:slim
```

Comment:
Probably dumb question. I see how it gets tagged as for example: `python-gitlab:v3.1.0-slim` but how does it get tagged as `python-gitlab:slim` ?

Reply:
My assumption is that line 29 here tags it as `python-gitlab:v3.1.0-slim`

Reply:
Ah. Line 15 is what does that tag. And line 29 is tagging it as `python-gitlab:slim`

Reply:
So kaniko with a `--destination` flag will automatically push to the registry with what's defined there that's where we tag it with the two `:${CI_COMMIT_TAG}-<flavor>` tags. After the first stage, we have these 2 in the registry.

So by the time we are calling `crane tag`, those version-tagged images are already in the registry and `crane tag` just re-applies these tags in this job remotely (more efficient than doing it locally).

kaniko docs:
https://github.com/GoogleContainerTools/kaniko

crane tag:
https://github.com/google/go-containerregistry/blob/main/cmd/crane/doc/crane_tag.md#crane-tag